server/controller/store.go

```diff
@@ -217,3 +185,235 @@ func (s storeCtrl) Delete(stream storev1.StoreService_DeleteServer) error {
+func (s storeCtrl) validateAndGetCID(record *corev1.Record) (string, error) {
+	if err := s.validateRecord(record); err != nil {
+		return "", err
+	}
+
+	// Calculate CID for the record
+	recordCID := record.GetCid()
+	if recordCID == "" {
+		storeLogger.Error("Failed to calculate record CID")
+
+		return "", status.Error(codes.Internal, "failed to calculate record CID")
+	}
+
+	storeLogger.Debug("CID calculated successfully", "cid", recordCID)
+
+	return recordCID, nil
+}
+
+// pushRecordToStore pushes a record to the store and adds it to the search index.
+func (s storeCtrl) pushRecordToStore(ctx context.Context, record *corev1.Record) (*corev1.RecordRef, error) {
+	// Push the record to store
+	pushedRef, err := s.store.Push(ctx, record)
+	if err != nil {
+		storeLogger.Error("Failed to push record to store", "error", err)
+
+		return nil, status.Errorf(codes.Internal, "failed to push record to store: %v", err)
+	}
+
+	storeLogger.Info("Record pushed to store successfully", "cid", pushedRef.GetCid())
+
+	// Add record to search index for discoverability
+	// Use the adapter pattern to convert corev1.Record to types.Record
+	recordAdapter := adapters.NewRecordAdapter(record)
+	if err := s.db.AddRecord(recordAdapter); err != nil {
+		// Log error but don't fail the push operation
+		storeLogger.Error("Failed to add record to search index", "error", err, "cid", pushedRef.GetCid())
+	} else {
+		storeLogger.Debug("Record added to search index successfully", "cid", pushedRef.GetCid())
+	}
+
+	return pushedRef, nil
+}
+
+// pushSignatureIfProvided pushes a signature artifact if the store supports it.
+func (s storeCtrl) pushSignatureToStore(ctx context.Context, signature *signv1.Signature, recordCID string) error {
+	storeLogger.Debug("Processing signature artifact")
+
+	// Check if store supports signature artifacts (OCI store should)
+	if ociStore, ok := s.store.(types.SignatureStoreAPI); ok {
+		err := ociStore.PushSignature(ctx, recordCID, signature)
+		if err != nil {
+			storeLogger.Error("Failed to push signature", "error", err, "cid", recordCID)
+
+			return status.Errorf(codes.Internal, "failed to push signature: %v", err)
+		}
+	} else {
+		storeLogger.Warn("Store does not support signature artifacts, ignoring", "storeType", fmt.Sprintf("%T", s.store))
+	}
+
+	return nil
+}
+
+// validateRecordRef validates a record reference.
+func (s storeCtrl) validateRecordRef(recordRef *corev1.RecordRef) error {
+	if recordRef.GetCid() == "" {
+		return status.Error(codes.InvalidArgument, "record cid is required")
+	}
+
+	return nil
+}
+
+// pullRecordFromStore pulls a record from the store with validation.
+func (s storeCtrl) pullRecordFromStore(ctx context.Context, recordRef *corev1.RecordRef) (*corev1.Record, error) {
+	// Check if record exists
+	_, err := s.store.Lookup(ctx, recordRef)
+	if err != nil {
+		st := status.Convert(err)
+
+		return nil, status.Errorf(st.Code(), "failed to lookup record: %s", st.Message())
+	}
```

Comment:
I'm wondering if we really need Lookup here. Both oci.Pull and oci.Lookup call the `fetchAndParseManifest` method, where we already do a Resolve.
server/store/oci/internal.go
```go
// Resolve manifest from remote tag (this also checks existence and validates CID format)
    manifestDesc, err := s.repo.Resolve(ctx, cid)
    
if err != nil {
        internalLogger.Debug("Failed to resolve manifest", "cid", cid, "error", err)
        return nil, nil, status.Errorf(codes.NotFound, "record not found: %s", cid)
    }
```
What do you think?

Reply:
you are absolutely right 👍 the `Pull` method in oci.go already looks up, thanks for catching this